[DOC] Doc updates for 2.2 release (#2740)
grafana/tempo: docs/sources/tempo/release-notes/v2-2.md

- Major additions to the TraceQL language: structural operators (descendant, child, sibling), results grouped by attribute (`by()`), and 3 new intrinsic attributes. 
- Faster TraceQL results, thanks to a streaming endpoint that returns partial results as the query executes as well as a multitude of performance-related improvements. 
- An experimental metrics-summary API that returns RED metrics for recently received spans grouped by your attribute of choice.

Tempo 2.2 makes vParquet2, a Parquet version designed to be more compatible with other Parquet implementations, the default block format. This block format is required for using structural operators and improves query performance relative to previous formats.

Read the [Tempo 2.2 blog post](/blog/NEED BLOG LINK) for more examples and details about these improvements.

{{% admonition type ="note" %}}
For a complete list of changes, enhancements, and bug fixes refer to the [Tempo 2.2 changelog](https://github.com/grafana/tempo/releases).
{{% /admonition %}}

## Features and enhancements

Some of the most important features and enhancements in Tempo 2.2 are highlighted below.

### Expanding the TraceQL language

> With this release, we’ve added to the [TraceQL language](https://grafana.com/docs/tempo/latest/traceql/). 

Author: ```suggestion
With this release, we’ve added to the [TraceQL language]({{< relref "../traceql/" >}}). 
```

{{% /admonition %}}

## Features and enhancements

Some of the most important features and enhancements in Tempo 2.2 are highlighted below.

### Expanding the TraceQL language

With this release, we’ve added to the [TraceQL language]({{< relref "../traceql" >}}).
TraceQL now offers:

* Structural operators: descendant (>>), child (>), and sibling (~) ([documentation]({{< relref "../traceql#structural" >}})). Find relevant traces based on their structure and relationships among spans. [PR [#2625](https://github.com/grafana/tempo/pull/2625) [#2660](https://github.com/grafana/tempo/pull/2660)]
* A `select()` operation that allows you to specify arbitrary span attributes that you want included in the TraceQL response ([documentation]({{< relref "../traceql#selection" >}})) [PR [2494](https://github.com/grafana/tempo/pull/2494)]
* A `by()` operation that groups span sets within a trace by an attribute of your choosing. This operation is not supported in the Grafana UI yet; you can only use `by()` when querying Tempo’s search API directly. ([documentation]({{< relref "../traceql#grouping" >}})  [PR  [2490](https://github.com/grafana/tempo/pull/2490)]

> * New intrinsic attributes for use in TraceQL queries: `traceDuration, `rootName`, and `rootServiceName` ([documentation]({{< relref "../traceql" >}})) [PR [#2503](https://github.com/grafana/tempo/pull/2503)]

Reviewer: ```suggestion
* New intrinsic attributes for use in TraceQL queries: `traceDuration`, `rootName`, and `rootServiceName` ([documentation]({{< relref "../traceql" >}})) [PR [#2503](https://github.com/grafana/tempo/pull/2503)]
```

> Read the [Tempo 2.2 blog post](/blog/need/link) for examples of how to use these new language additions. \

Reviewer: ```suggestion
Read the [Tempo 2.2 blog post](/blog/need/link) for examples of how to use these new language additions. 
```

To learn more about the TraceQL syntax, see the [TraceQL documentation]({{< relref "../traceql" >}}).
For information on planned future extensions to the TraceQL language, see [future work]({{< relref "../traceql/architecture" >}}).

### Get TraceQL results faster

We’re always trying to reduce the time you spend waiting to get results to your TraceQL queries, and we’ve made some nice progress on this front with this release.

> We’ve added a [GRPC streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc)endpoint to Tempo’s query frontend that allows a client to stream search results from Tempo. The Tempo CLI [PR  [#2366](https://github.com/grafana/tempo/pull/2366)] has been updated to use this new streaming endpoint; As of version 10.1, Grafana supports it as well, though you must first enable the `traceQLStreaming` feature toggle [PR [#72288](https://github.com/grafana/grafana/pull/72288)].

Reviewer: ```suggestion
We’ve added a [GRPC streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc) endpoint to Tempo’s query frontend that allows a client to stream search results from Tempo. The Tempo CLI [PR  [#2366](https://github.com/grafana/tempo/pull/2366)] has been updated to use this new streaming endpoint; As of version 10.1, Grafana supports it as well, though you must first enable the `traceQLStreaming` feature toggle [PR [#72288](https://github.com/grafana/grafana/pull/72288)].
```

From here, you might see that spans from `namespace=A` have a significantly higher error rate than those from `namespace=B`.
As another example, you could use this API to compare latencies of your spans broken down by the `region` attribute.
From here, you might notice that spans from `region=North-America` have higher latencies than those from `region=Asia-Pacific`.

This API is meant to enable ad-hoc analysis of your incoming spans; by segmenting your spans by attribute and looking for differences in RED metrics, you can more quickly isolate where problems like elevated error rates or higher latencies are coming from.

Unlike RED metrics computed by Tempo’s metrics-generator, the values returned by this API are not persisted as time series. This has the advantage that you do not need to provide your own time series databases for storing and querying these metrics. It also allows you to compute RED metrics broken down by high cardinality attributes that would be too expensive to store in a time series database. Use the metrics generator if you want to store and visualize RED metrics over multi-hour or multi-day time ranges, or you want to alert on these metrics.

> To learn more about this API, refer to the[ metrics summary API documentation.](/docs/grafana-cloud/data-configuration/traces/metrics-summary-api/)

Reviewer: does this need updated to point to a tempo page if you're saying this has been moved out of cloud docs?

This work is represented in multiple PRs: [2368](https://github.com/grafana/tempo/pull/2368), [2418](https://github.com/grafana/tempo/pull/2418), [2424](https://github.com/grafana/tempo/pull/2424), [2442](https://github.com/grafana/tempo/pull/2442), [2480](https://github.com/grafana/tempo/pull/2480), [2481](https://github.com/grafana/tempo/pull/2481), [2501](https://github.com/grafana/tempo/pull/2501), [2579](https://github.com/grafana/tempo/pull/2579), and [2582](https://github.com/grafana/tempo/pull/2582).

### Other enhancements

> * Tempo’s [tag values]({{< relref "../api_docs#search-tag-values" >}}) and [tag names]({{< relref "../api_docs#search-tags" >}}) APIs now support filtering [PR [2253](https://github.com/grafana/tempo/pull/2253)]. This lets you retrieve all valid attribute values and names given certain criteria. For example, you can get a list of values for the attribute `namespace` seen on spans with attribute `resource=A.` This feature is off by default; to enable, configure `autocomplete_filtering_enabled`. ([documentation]({{< relref "../api_docs" >}})). Grafana’s autocomplete can make use of this filtering capability to provide better suggestions starting in v10.2 [[PR 67845]](https://github.com/grafana/grafana/pull/67845).

Reviewer: ```suggestion
* Tempo’s [tag values]({{< relref "../api_docs#search-tag-values" >}}) and [tag names]({{< relref "../api_docs#search-tags" >}}) APIs now support filtering [PR [2253](https://github.com/grafana/tempo/pull/2253)]. This lets you retrieve all valid attribute values and names given certain criteria. For example, you can get a list of values for the attribute `namespace` seen on spans with attribute `resource=A.` This feature is off by default; to enable, configure `autocomplete_filtering_enabled`. ([documentation]({{< relref "../api_docs" >}})). Grafana’s autocomplete can make use of this filtering capability to provide better suggestions starting in v10.2 [PR [67845]](https://github.com/grafana/grafana/pull/67845).
```

* Tempo’s metrics-generator now supports span filtering. Setting up filters allows you to compute metrics over the specific spans you care about, excluding others. It also can reduce the cardinality of generated metrics, and therefore the cost of storing those metrics in a Prometheus-compatible TSDB. ([documentation]({{< relref "../metrics-generator/span_metrics#filtering" >}})) [PR [2274](https://github.com/grafana/tempo/pull/2274)]

* Tempo’s metrics-generator can now detect virtual nodes ([documentation]({{< relref "../metrics-generator/service_graphs#virtual-nodes" >}})) [PR [2365](https://github.com/grafana/tempo/pull/2365)]. As a result, you’ll now see these virtual nodes represented in your service graph. For more information, refer to the [virtual nodes documentation]({{< relref "../metrics-generator/service_graphs#virtual-nodes" >}}).

## Upgrade considerations

When [upgrading]({{< relref "../setup/upgrade" >}}) to Tempo 2.2, be aware of these breaking changes:

> * JSonnet users only: Convert metrics generator from deployment to a statefulset in JSonnet. Refer to the PR for seamless migration instructions. [PRs [#2533](https://github.com/grafana/tempo/pull/2533), [#2467](https://github.com/grafana/tempo/pull/2647)]

Reviewer: ```suggestion
* JSonnet users only: We've converted the metrics-generator component from a k8s deployment to a k8s statefulset. Refer to the PR for seamless migration instructions. [PRs [#2533](https://github.com/grafana/tempo/pull/2533), [#2467](https://github.com/grafana/tempo/pull/2647)]
```

<p>
<code>  ```</code>
   </td>
   <td>Remove support for `tolerant_failed_blocks` [PR <a href="https://github.com/grafana/tempo/pull/2416">2416</a>]
   </td>
  </tr>
  <tr>
   <td><code>```</code>
<p>
<code>storage:</code>
<p>
<code>  trace:</code>
<p>
<code>    s3:</code>
<p>
<code>      insecure_skip_verify: true   // renamed to tls_insecure_skip_verify</code>
<p>
<code>```</code>
   </td>
   <td>Renamed `insecure_skip_verify` to `tls_insecure_skip_verify` [PR <a href="https://github.com/grafana/tempo/pull/2407">2407</a>]
   </td>
  </tr>
</table>

## Bug fixes

For a complete list, refer to the [Tempo changelog](https://github.com/grafana/tempo/releases).

* Fixed an issue in the metrics-generator that prevented scaling up parallelism when remote writing of metrics was lagging behind [PR [2463](https://github.com/grafana/tempo/issues/2463)]
* Fixed an issue where metrics-generator was setting wrong labels for `traces_target_info` [PR [2546](https://github.com/grafana/tempo/pull/2546)]
* Fixed an issue where matches and other spanset level attributes were not persisted to the TraceQL results. [PR [2490](https://github.com/grafana/tempo/pull/2490)]

> * Fixed an issue where ingester search could occasionally fail with ‘file does not exist’ error [PR [2534](https://github.com/grafana/tempo/issues/2534)]

Reviewer: ```suggestion
* Fixed an issue where ingester search could occasionally fail with `file does not exist` error [PR [2534](https://github.com/grafana/tempo/issues/2534)]
```

---
title: Version 2.2 release notes
menuTitle: V2.2
description: Release notes for Grafana Tempo 2.2
weight: 50
---

# Version 2.2 release notes

The Tempo team is pleased to announce the release of Tempo 2.2.

This release gives you:

- Major additions to the TraceQL language: structural operators (descendant, child, sibling), results grouped by attribute (`by()`), and 3 new intrinsic attributes.
- Faster TraceQL results, thanks to a streaming endpoint that returns partial results as the query executes as well as a multitude of performance-related improvements.
- An experimental metrics-summary API that returns RED metrics for recently received spans grouped by your attribute of choice.

Tempo 2.2 makes vParquet2, a Parquet version designed to be more compatible with other Parquet implementations, the default block format. This block format is required for using structural operators and improves query performance relative to previous formats.

> Read the [Tempo 2.2 blog post](/blog/NEED BLOG LINK) for more examples and details about these improvements.

Reviewer 2: :chicken: and :egg: here. i need this backported to 2.2 so i can link to it from the blog. i think we should get this in the 2.2 release branch first then we can go back and add the correct blog link later

Author: I can give you the link for the rel notes: https://grafana.com/docs/tempo/latest/release-notes/v2-2/

Reviewer 2: added to blogpost!

{{% admonition type ="note" %}}

> For a complete list of changes, enhancements, and bug fixes refer to the [Tempo 2.2 changelog](https://github.com/grafana/tempo/releases).

Reviewer 2: ```suggestion
For a complete list of changes, enhancements, and bug fixes refer to the [Tempo 2.2 changelog](https://github.com/grafana/tempo/releases/tag/v2.2.0).
```

{{% /admonition %}}

## Features and enhancements

Some of the most important features and enhancements in Tempo 2.2 are highlighted below.

### Expanding the TraceQL language

> With this release, we’ve added to the [TraceQL language]({{<relref "../traceql" >}}).

Author: ```suggestion
With this release, we’ve added to the [TraceQL language]({{< relref "../traceql" >}}).
```

* Structural operators: descendant (>>), child (>), and sibling (~) ([documentation]({{< relref "../traceql#structural" >}})). Find relevant traces based on their structure and relationships among spans. [PR [#2625](https://github.com/grafana/tempo/pull/2625) [#2660](https://github.com/grafana/tempo/pull/2660)]
* A `select()` operation that allows you to specify arbitrary span attributes that you want included in the TraceQL response ([documentation]({{< relref "../traceql#selection" >}})) [PR [2494](https://github.com/grafana/tempo/pull/2494)]
* A `by()` operation that groups span sets within a trace by an attribute of your choosing. This operation is not supported in the Grafana UI yet; you can only use `by()` when querying Tempo’s search API directly. ([documentation]({{< relref "../traceql#grouping" >}})  [PR  [2490](https://github.com/grafana/tempo/pull/2490)]
* New intrinsic attributes for use in TraceQL queries: `traceDuration, `rootName`, and `rootServiceName` ([documentation]({{< relref "../traceql" >}})) [PR [#2503](https://github.com/grafana/tempo/pull/2503)]

Read the [Tempo 2.2 blog post](/blog/need/link) for examples of how to use these new language additions. \

> To learn more about the TraceQL syntax, see the [TraceQL documentation](http://localhost:3002/docs/tempo/latest/traceql/).

Reviewer 2: localhost links?

Author: Thank you. Damn. Fixing that.

> For information on planned future extensions to the TraceQL language, see [future work](http://localhost:3002/docs/tempo/latest/traceql/architecture/).

Author: ```suggestion
For information on planned future extensions to the TraceQL language, see [future work]({{< relref "../traceql/architecture" >}}).
```

### Get TraceQL results faster

We’re always trying to reduce the time you spend waiting to get results to your TraceQL queries, and we’ve made some nice progress on this front with this release.

> We’ve added a [GRPC streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc)endpoint to Tempo’s query frontend that allows a client to stream search results from Tempo. The Tempo CLI [PR  [#2366](https://github.com/grafana/tempo/pull/2366)] has been updated to use this new streaming endpoint; As of version 10.1, Grafana uses it as well [PR [#67591](https://github.com/grafana/grafana/pull/69212)].

Reviewer: ```suggestion
We’ve added a [GRPC streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc)endpoint to Tempo’s query frontend that allows a client to stream search results from Tempo. The Tempo CLI [PR  [#2366](https://github.com/grafana/tempo/pull/2366)] has been updated to use this new streaming endpoint; As of version 10.1, Grafana supports it as well, though you must first enable the `traceQLStreaming` feature toggle [PR [#72288](https://github.com/grafana/grafana/pull/72288)].
```

* Tempo’s metrics-generator now supports span filtering. Setting up filters allows you to compute metrics over the specific spans you care about, excluding others. It also can reduce the cardinality of generated metrics, and therefore the cost of storing those metrics in a Prometheus-compatible TSDB. ([documentation]({{< relref "../metrics-generator/span_metrics#filtering" >}})) [PR [2274](https://github.com/grafana/tempo/pull/2274)]

* Tempo’s metrics-generator can now detect virtual nodes ([documentation]({{< relref "../metrics-generator/service_graphs#virtual-nodes" >}})) [PR [2365](https://github.com/grafana/tempo/pull/2365)]. As a result, you’ll now see these virtual nodes represented in your service graph. For more information, refer to the [virtual nodes documentation]({{< relref "../metrics-generator/service_graphs#virtual-nodes" >}}).

## Upgrade considerations

> When [upgrading](http://localhost:3002/docs/tempo/latest/setup/upgrade/) to Tempo 2.2, be aware of these breaking changes:

Reviewer 2: localhost

Author: ```suggestion
When [upgrading]({{< relref "../setup/upgrade" >}}) to Tempo 2.2, be aware of these breaking changes:
```

> * Convert metrics generator from deployment to a statefulset in JSonnet. Refer to the PR for seamless migration instructions. [PRs [#2533](https://github.com/grafana/tempo/pull/2533), [#2467](https://github.com/grafana/tempo/pull/2647)]

Reviewer 2: can we just add a quick note "Impacts jsonnet users only" or something?

Author: ```suggestion
* JSonnet users only: Convert metrics generator from deployment to a statefulset in JSonnet. Refer to the PR for seamless migration instructions. [PRs [#2533](https://github.com/grafana/tempo/pull/2533), [#2467](https://github.com/grafana/tempo/pull/2647)]
```

   </td>
  </tr>
  <tr>
   <td><code>```</code>
<p>
<code>storage:</code>
<p>
<code>  trace:</code>
<p>
<code>    s3:</code>
<p>
<code>      insecure_skip_verify: true   // renamed to tls_insecure_skip_verify</code>
<p>
<code>```</code>
   </td>
   <td>Renamed `insecure_skip_verify` to `tls_insecure_skip_verify` [PR <a href="https://github.com/grafana/tempo/pull/2407">2407</a>]
   </td>
  </tr>
</table>

## Bug fixes

For a complete list, refer to the [Tempo changelog](https://github.com/grafana/tempo/releases).

* Fixed an issue in the metrics-generator that prevented scaling up parallelism when remote writing of metrics was lagging behind [PR [2463](https://github.com/grafana/tempo/issues/2463)]
* Fixed an issue where metrics-generator was setting wrong labels for `traces_target_info` [PR [2546](https://github.com/grafana/tempo/pull/2546)]
* Fixed an issue where matches and other spanset level attributes were not persisted to the TraceQL results. [PR [2490](https://github.com/grafana/tempo/pull/2490)]
* Fixed an issue where ingester search could occasionally fail with ‘file does not exist’ error [PR [2534](https://github.com/grafana/tempo/issues/2534)]

> * Fixed an issue where Tempo failed to find the `meta.json` path after adding a prefix in object storage configuration. [PR [2585](https://github.com/grafana/tempo/issues/2585)]

Author: Removing from here since it was removed in the draft rel notes. 
```suggestion
```
